numpy/core/tests/data/generate_umath_validation_data.cpp

```diff
@@ -0,0 +1,157 @@
+#define MINFLT std::numeric_limits<T>::min()
+#define MAXFLT std::numeric_limits<T>::max()
+#define INF    std::numeric_limits<T>::infinity()
+#define qNAN   std::numeric_limits<T>::quiet_NaN()
+#define sNAN   std::numeric_limits<T>::signaling_NaN()
+
+template<typename T>
+std::vector<T> generate_input_vector(std::string func) {
+    std::vector<T> input = {MINDEN, -MINDEN, MINFLT, -MINFLT, MAXFLT, -MAXFLT,
+                            INF, -INF, qNAN, sNAN, -1.0, 1.0, 0.0, -0.0};
+
+    // [-1.0, 1.0]
+    if ((func == "arcsin") || (func == "arccos") || (func == "arctanh")){
+        append_random_array<T>(input, -1.0, 1.0, 700);
+    }
+    // (0.0, INF]
+    else if ((func == "log2") || (func == "log10")) {
+        append_random_array<T>(input, 0.0, 1.0, 200);
+        append_random_array<T>(input, MINDEN, MINFLT, 200);
+        append_random_array<T>(input, MINFLT, 1.0, 200);
+        append_random_array<T>(input, 1.0, MAXFLT, 200);
+    }
+    // (-1.0, INF]
+    else if (func == "log1p") {
+        append_random_array<T>(input, -1.0, 1.0, 200);
+        append_random_array<T>(input, -MINFLT, -MINDEN, 100);
+        append_random_array<T>(input, -1.0, -MINFLT, 100);
+        append_random_array<T>(input, MINDEN, MINFLT, 100);
+        append_random_array<T>(input, MINFLT, 1.0, 100);
+        append_random_array<T>(input, 1.0, MAXFLT, 100);
+    }
+    // [1.0, INF]
+    else if (func == "arccosh") {
+        append_random_array<T>(input, 1.0, 2.0, 400);
+        append_random_array<T>(input, 2.0, MAXFLT, 300);
+    }
+    // [-INF, INF]
+    else {
+        append_random_array<T>(input, -1.0, 1.0, 100);
+        append_random_array<T>(input, MINDEN, MINFLT, 100);
+        append_random_array<T>(input, -MINFLT, -MINDEN, 100);
+        append_random_array<T>(input, MINFLT, 1.0, 100);
+        append_random_array<T>(input, -1.0, -MINFLT, 100);
+        append_random_array<T>(input, 1.0, MAXFLT, 100);
+        append_random_array<T>(input, -MAXFLT, -100.0, 100);
+    }
+
+    std::random_shuffle(input.begin(), input.end());
+    return input;
+}
+
+int main() {
+    srand (42);
+    std::vector<struct ufunc> umathfunc = {
+        {"sin",sin,sin,2.37,3.3},
+        {"cos",cos,cos,2.36,3.38},
+        {"tan",tan,tan,3.91,3.93},
+        {"arcsin",asin,asin,3.12,2.55},
+        {"arccos",acos,acos,2.1,1.67},
+        {"arctan",atan,atan,2.3,2.52},
+        {"sinh",sinh,sinh,1.55,1.89},
+        {"cosh",cosh,cosh,2.48,1.97},
+        {"tanh",tanh,tanh,1.38,1.19},
+        {"arcsinh",asinh,asinh,1.01,1.48},
+        {"arccosh",acosh,acosh,1.16,1.05},
+        {"arctanh",atanh,atanh,1.45,1.46},
+        {"cbrt",cbrt,cbrt,1.94,1.82},
+        //{"exp",exp,exp,3.76,1.53},
+        {"exp2",exp2,exp2,1.01,1.04},
+        {"expm1",expm1,expm1,2.62,2.1},
+        //{"log",log,log,1.84,1.67},
+        {"log10",log10,log10,3.5,1.92},
+        {"log1p",log1p,log1p,1.96,1.93},
+        {"log2",log2,log2,2.12,1.84},
```

Comment:
I am confused about the f32ulp and f64ulp values here (maybe move the struct definition closer to this code so it is clear what the fields are). How did you calculate them as floats?

Reply:
Cleared up on a community call: these are actually the difference between the values so they should be floating point.